### PR TITLE
SSH tunnel properties

### PR DESCRIPTION
**Pull request by ankitsiam**
Defines SDK properties required for SSH Tunneling functionality in runtime.

### CODE REVIEW

File: lib/collection/sshtest.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+var SSHTunnell = require('./ssh-tunnell').SSHTunnell,
```
Comment (kunagpal) [inline lib/collection/sshtest.js]: Single `l`

File: lib/collection/sshtest.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    SSHAuth = require('./ssh-auth').SSHAuth,
+    tunnell = new SSHTunnell(),
+
+    config = new SSHConfig({
+        forwardPort: 3128,
+        auth: new SSHAuth({
+            host: '54.173.108.141',
+            port: 22,
+            username: 'ubuntu',
+            privateKeyPath: '/Users/ankitsingh/Documents/devPlayground/ssh/key_rsa'
```
Comment (kunagpal) [inline lib/collection/sshtest.js]: 🤦‍♂️

File: lib/collection/sshtest.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        })
+    });
+
+// console.log(config.auth.privateKeyPath);
```
Comment (kunagpal) [inline lib/collection/sshtest.js]: Remove.

File: lib/collection/sshtest.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+tunnell.start(config, 8000, function (err) {
+    if (err) {
+        // throw err;
+        console.log('somthing is wrong');
```
Comment (kunagpal) [inline lib/collection/sshtest.js]: Log the actual error. `Something is/went wrong` messages are seldom useful.

File: lib/collection/sshtest.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    console.log('SSH tunnell established succesfully');
+});
+
+setTimeout(function () {
+    tunnell.close(function (err, status) {
+        if (err) {
+            console.log(err);
+            return;
+        }
+        console.log('Status : ' + status);
```
Comment (kunagpal) [inline lib/collection/sshtest.js]: If the status is always successful when the `err` is `null`, this log can be removed.

File: lib/collection/ssh-config.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+             * @type {SSHAuth}
+             */
+            auth: new SSHAuth(),
+
+            /**
+             * How often (in milliseconds) to send SSH-level keepalive packets to the server, set to 0 to disable.
+             * Default: 0
+             * @type {Integer}
+             */
+            keepaliveInterval: DEFAULT_KEEP_ALIVE_INTERVEL,
```
Comment (kunagpal) [inline lib/collection/ssh-config.js]: I like how the same word can be spelt in two different ways in the same line 😛

File: lib/collection/ssh-config.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+             * disconnection
+             * Default : 3
+             * @type {Integer}
+             */
+            keepaliveCountMax: DEFAULT_KEEP_ALIVE_COUNT_MAX,
+
+            /**
+             * How long (in milliseconds) to wait for the SSH handshake to complete.
+             * Default :  20000
+             * @type {Integer}
```
Comment (kunagpal) [inline lib/collection/ssh-config.js]: Use the `@default` tag here.

File: lib/collection/ssh-config.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+     * Updates the proerties of SSHConfig object based on the options provided
+     * @param {SSHConfig~definition} options
+     */
+
+    update: function (options) {
+        if (!_.isObject(options)) {
+            return;
+        }
+
+        var forwardPort = _.get(options, 'forwardPort') >> 0,
```
Comment (kunagpal) [inline lib/collection/ssh-config.js]: Minimize the use of `_.get` here.

File: lib/collection/ssh-config.js [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            keepaliveCountMax = _.get(options, 'keepaliveCountMax'),
+            readyTimeout = _.get(options, 'readyTimeout');
+
+        _.isBoolean(options.enablePortForwarding) && (this.enablePortForwarding = options.enablePortForwarding);
+        SSHAuth.isSSHAuth(options.auth) && (this.auth.update(options.auth));
+        forwardPort && (this.forwardPort = options.forwardPort);
+        (keepaliveInterval >> 0 || keepaliveInterval === 0) && (this.keepaliveInterval = options.keepaliveInterval);
+        (keepaliveCountMax >> 0 || keepaliveCountMax === 0) && (this.keepaliveCountMax = options.keepaliveCountMax);
+        (readyTimeout >> 0 || readyTimeout === 0) && (this.readyTimeout = options.readyTimeout);
+
```
Comment (kunagpal) [inline lib/collection/ssh-config.js]: Extra blank line.

File: lib/collection/ssh-auth.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+    update: function (authOptions) {
+
+        if (!_.isObject(authOptions)) {
+            return;
+        }
+
+        var port = _.get(authOptions, 'port') >> 0;
+
+        port && (this.port = authOptions.port);
```
[the remark this answers is not in the document]
Comment (ankitsiam) [reply lib/collection/ssh-auth.js]: Okay, what exactly do you mean ?

File: lib/collection/ssh-auth.js [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+     * @readOnly
+     * @type {Boolean}
+     */
+
+    _postman_propertyRequiresId: true,
+
+    /**
+     * Updates the properties of the proxy object based on the options provided.
+     *
+     * @param {ProxyConfig~definition} options The proxy object structure.
```
Comment (codenirvana) [inline lib/collection/ssh-auth.js]: Its not `ProxyConfig` class

File: lib/collection/ssh-auth.js [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+     */
+
+    update: function (authOptions) {
+
```
Comment (codenirvana) [inline lib/collection/ssh-auth.js]: Extra newline

File: lib/collection/ssh-auth.js (hunk 11 above)
Comment (codenirvana) [reply lib/collection/ssh-auth.js]: Missing JSDoc

File: lib/collection/ssh-auth.js [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+             * Username of the user one want to login as
+             * @type {String}
+             */
+            username: E,
+
+            /**
+             * Password of the user one want to login as
+             * @type {String}
+             */
+            userPassword: E,
```
Comment (codenirvana) [inline lib/collection/ssh-auth.js]: Why not just`password`?

File: lib/collection/ssh-auth.js [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+), Property);
+
+_.assign(SshAuth.prototype, /** @lends SshAuth.prototype */ {
+    /**
+     * Defines whether this property instances requires an id
+     * @private
+     * @readOnly
+     * @type {Boolean}
+     */
+
```
Comment (codenirvana) [inline lib/collection/ssh-auth.js]: Extra newline

File: lib/collection/ssh-auth.js [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+     */
+    _postman_propertyName: 'SshAuth',
+
+    /**
+     * Check whether an object is an instance of PostmanItem.
+     *
+     * @param {*} obj
+     * @returns {Boolean}
+     */
+
```
Comment (codenirvana) [inline lib/collection/ssh-auth.js]: Extra newline

File: lib/collection/ssh-config.js [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+_.inherit((
+
+    /**
+     * A SSH config definition which represents the configuration for a SSH Tunnel.
+     * @constructor
+     * @extends {Property}
+     * @param {SshConfig~definition=} [options] - sepcifies object containing info required to establish a ssh tunnel
+    */
+
```
Comment (codenirvana) [inline lib/collection/ssh-config.js]: Extra newline

File: lib/collection/ssh-config.js [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+             * @type {SshAuth}
+             */
+            auth: new SshAuth(),
+
+            /**
+             * How often (in milliseconds) to send SSH-level keepalive packets to the server, set to 0 to disable.
+             * Default: 0
+             * @type {Integer}
+             */
+            keepaliveIntervel: DEFAULT_KEEP_ALIVE_INTERVEL,
```
Comment (codenirvana) [inline lib/collection/ssh-config.js]: `keepAliveInterval` ?

File: lib/collection/ssh-config.js [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+     * @readOnly
+     * @type {Boolean}
+     */
+    _postman_propertyRequiresId: true,
+
+    /**
+     * Updates the proerties of SshConfig object based on the options provided
+     * @param {SshConfig~definition} options
+     */
+
```
Comment (codenirvana) [inline lib/collection/ssh-config.js]: Extra newline

Comment (codenirvana) [reply lib/collection/ssh-config.js]: Missing JSDoc

File: lib/collection/ssh-config.js [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+    update: function (options) {
+        if (!_.isObject(options)) {
+            return;
+        }
+
+        var forwardPort = options.forwardPort,
+            keepaliveIntervel = options.keepaliveIntervel,
+            keepaliveCountMax = options.keepaliveCountMax,
```
Comment (codenirvana) [inline lib/collection/ssh-config.js]: `keepAlive`

File: lib/collection/ssh-config.js [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+            readyTimeOut = options.readyTimeOut;
+
+        _.isObject(options.auth) && (SshAuth.isSshAuth(options.auth)) ?
```
Comment (kunagpal) [inline lib/collection/ssh-config.js]:
```js
var auth = options.auth;

this.auth.update(SshAuth.isSshAuth(auth) ? auth : new SshAuth(auth)); 
```

File: lib/collection/ssh-config.js [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+            (this.auth.update(options.auth)) : (this.auth.update(new SshAuth(options.auth)));
+
+        _.isBoolean(options.enablePortForwarding) && (this.enablePortForwarding = options.enablePortForwarding);
+        forwardPort && (this.forwardPort = options.forwardPort);
+        (keepaliveIntervel >> 0 || keepaliveIntervel === 0) && (this.keepaliveIntervel = options.keepaliveIntervel);
```
Comment (kunagpal) [inline lib/collection/ssh-config.js]: All validation deferred to runtime.

File: lib/collection/ssh-config.js (hunk 16 above)
Comment (ankitsiam) [reply lib/collection/ssh-config.js]: as per esLint rules

Comment (codenirvana) [reply lib/collection/ssh-config.js]: Which rule?

File: lib/collection/ssh-auth.js [hunk 22]
```diff
@@ -0,0 +1,152 @@
+var _ = require('../util').lodash,
+    Property = require('./property').Property,
+    E = '',
+    DEFAULT_SSH_PORT = 22,
+
+    SshAuth;
+
+/**
+* The following is the object structure accepted as constructor parameter while calling `new SshAuth(...)`. It is
```
Comment (shamasis) [inline lib/collection/ssh-auth.js]: the `*` of these comments are missing one leading whitespace.

File: lib/collection/ssh-auth.js [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+     *     mySshAuth = new SshAuth({
+     *          host: "54.111.103.121",
+     *          port: "22"
+     *          username: "someuser",
+     *          password: "somepassword",
+     *          privateKeyPath: "/path/to/key/file",
+     *          passPhrase: "somepassphrase"
+     *      });
+     */
+    SshAuth = function SshAuth (definition) {
```
Comment (shamasis) [inline lib/collection/ssh-auth.js]: `SSHAuth`

Comment (ankitsiam) [reply lib/collection/ssh-auth.js]: Actually, there's a test in postman-collection which prevents this. Basically, test enforces that first character should be Uppercase and rest lowercase.

**Pull request closed without merging**